src/baldr/shortcut_recovery.h

```diff
@@ -0,0 +1,231 @@
+#pragma once
+
+#include "baldr/graphreader.h"
+#include "baldr/tilehierarchy.h"
+#include "midgard/logging.h"
+
+#include <cstdint>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace {
+// TODO: break this out into a private header
+// a static cache for shortcut recovery that we can optionally pre-fill
+struct shortcut_recovery_t {
+protected:
+  /**
+   * Constructs a shortcut cache from a graphreaders tileset by recovering all shortcuts. If the
+   * graphreader passed in is null nothing is cached and revoery will happen on the fly
+   * @param reader
+   */
+  shortcut_recovery_t(valhalla::baldr::GraphReader* reader) {
```

Comment:
on startup (when we first construct a graph reader) this gets called. if you wanted to turn caching on it will run through your tileset and cache all the superceded edges of all the shortcuts, if you didnt it will just bail immediately. in both cases it logs about what it did.